### ISSUE TITLE
Distillation

### ISSUE THREAD

**Issue by JinSeoungwoo** [erfanzar/EasyDeL#44]
I would like to try a distillation where I calculate the softmax prob of the logits of the teacher model using my input dataset, save it, and use it as a soft target for the student model to train. Is this possible enough in easydel?

**Comment by JinSeoungwoo** [comment]
I also wondered if there was a way to use packing to maximize the efficiency of the training process. After [a little research](https://www.graphcore.ai/posts/packedbert-how-to-accelerate-nlp-tasks-for-transformers-with-packing) , I learned about using attention masks to minimize padding and combine multiple training sentences to train simultaneously.

**Comment by erfanzar** [comment]
yes actually i would like to do that too

**Comment by erfanzar** [comment]
right now im working on flash attn 2 in jax but it seems like it not working anygood in jax

**Comment by erfanzar** [comment]
i have tried 4 different ways but still at the end jax normal attention perform better than my flash attention s 
im thinking of using triton for that and create a bridge between jax and triron do you have any idea for that?
[check this out](https://github.com/erfanzar/FJFormer/tree/main/fjformer/attention)

**Comment by JinSeoungwoo** [comment]
> i have tried 4 different ways but still at the end jax normal attention perform better than my flash attention s im thinking of using triton for that and create a bridge between jax and triron do you have any idea for that? [check this out](https://github.com/erfanzar/FJFormer/tree/main/fjformer/attention)

I'll try to think of a way to implement flash attention 2. By the way, is there any code that can test attention?

**Comment by erfanzar** [comment]
im using triton to find a way to port that in jax but if you want to test the ones i have already implemented use fjformer implementation 

and for those func that i want to be tested only in jax i use this code

```python
def flash_attn_run(seq_len):
  q = random.normal(qs, (1, 32, seq_len, 64))
  k = random.normal(ks, (1, 32, seq_len, 64))
  v = random.normal(vs, (1, 32, seq_len, 64))
  res = efficient_attention(query=q, key=k, value=v, key_chunk_size=512, query_chunk_size=512)


def flax_attn_run(seq_len):
  q = random.normal(qs, (1, 32, seq_len, 64))
  k = random.normal(ks, (1, 32, seq_len, 64))
  v = random.normal(vs, (1, 32, seq_len, 64))
  res = flax_dot_product_attention(query=q, key=k, value=v)

sequence_length = [512, 1024, 2048, 4096, 8192, 16384, 32768]

perfplot.show(
    setup=lambda n: n,
    kernels=[flash_attn_run, flax_attn_run],
    labels=["flash_attn_run","flax_attn_run"],
    n_range=sequence_length,
    xlabel="sequence_length",
    equality_check=None,
    logx=True,
    logy=True,
    title="Customized Benchmark Plot"
)
```

and for efficiency test i use 
```python

dtype = jnp.float32

def reference_attn(query, key, value):

    query = query / jnp.sqrt(query.shape[-1]).astype(dtype)
    logits = jnp.einsum("bqhc,bkhc->bhqk", query, key)
    mask_value = jnp.finfo(logits.dtype).min
    _, q_seq_len, _, _ = query.shape
    _, kv_seq_len, _, _ = key.shape
    mask_shape = (q_seq_len, kv_seq_len)
    row_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 0)
    col_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 1)
    causal_mask = (row_ids < col_ids)[None, None, :, :]
    logits = logits + jnp.where(causal_mask, mask_value, 0.0)
    weights = jax.nn.softmax(logits, axis=-1)
    out = jnp.einsum("bhqk,bkhc->bqhc", weights, value)
    return out

def random_qkv(seq_len):
  q = random.normal(qs, (1, 32, seq_len, 64))
  k = random.normal(ks, (1, 32, seq_len, 64))
  v = random.normal(vs, (1, 32, seq_len, 64))
  return q,k,v


@partial(jax.jit, static_argnums=(0,))
def reference_attn_forward_backward( seq_len):
    @partial(jax.grad, argnums=(0, 1, 2))
    @partial(jax.checkpoint, policy=jax.checkpoint_policies.nothing_saveable())
    def grad_fn(query, key, value):
        out = reference_attn(query, key, value)
        return jnp.mean(out)

    query, key, value = random_qkv(seq_len)
    return jax.flatten_util.ravel_pytree(
        grad_fn(query, key, value)[1]
    )[0].mean()

@partial(jax.jit, static_argnums=(0,))
def efficient_attn_forward_backward(seq_len):
    @partial(jax.grad, argnums=(0, 1, 2))
    def grad_fn(query, key, value):
        out = efficient_attention(query, key, value,query_chunk_size=4,key_chunk_size=4)
        return jnp.mean(out)

    query, key, value = random_qkv(seq_len)
    return jax.flatten_util.ravel_pytree(
        grad_fn(query, key, value)[1]
    )[0].mean()

warmup_steps = 10
steps = 60


jax.block_until_ready(reference_attn_forward_backward(8192))
jax.block_until_ready(efficient_attn_forward_backward(8192))

all_results = []
for i in range(warmup_steps):
    all_results.append(reference_attn_forward_backward(8192))
jax.block_until_ready(all_results)

start_time = time.time()
all_results = []
for i in range(steps):
    all_results.append(reference_attn_forward_backward(8192))

jax.block_until_ready(all_results)
elapsed_time_ref_attn = time.time() - start_time
print(f'Reference attention: {elapsed_time_ref_attn:.3f} seconds')


all_results = []
for i in range(warmup_steps):
    all_results.append(efficient_attn_forward_backward(8192))
jax.block_until_ready(all_results)


start_time = time.time()
all_results = []
for i in range(steps):
    all_results.append(efficient_attn_forward_backward(8192))

jax.block_until_ready(all_results)
elapsed_time_efficient_attn = time.time() - start_time
print(f'Efficient attention: {elapsed_time_efficient_attn:.3f} seconds')

flops_ratio = (1) ** 2
efficiency = elapsed_time_ref_attn / elapsed_time_efficient_attn * flops_ratio
print(f'Efficiency: {efficiency:.3f}')
```

**Comment by JinSeoungwoo** [comment]
Thank you for providing the script. However, I don't know the values of qs,ks,vs, could you please provide them?

**Comment by JinSeoungwoo** [comment]
Can you tell me where in the training process it goes about predicting the next token? I don't know if this is correct, but I'm curious about the part where it predicts 4444 in [1, 22, 333] -> [1, 22, 333, 4444].

**Comment by erfanzar** [comment]
it's [here](https://github.com/erfanzar/EasyDeL/blob/main/lib/python/EasyDel/trainer/fsdp_train.py#L292) where the next token be predicted in training process with shape of (batch,seqlen,hidden_size)

**Comment by JinSeoungwoo** [comment]
> it's [here](https://github.com/erfanzar/EasyDeL/blob/main/lib/python/EasyDel/trainer/fsdp_train.py#L292) where the next token be predicted in training process with shape of (batch,seqlen,hidden_size)

Thanks for letting me know. I wanted to move the attention mask for each token and loss calculation for the packing mentioned above.

For example, 
if we had two sequences together like
[12,34,56,78,[EOS],87,65,43,21]

first make prediction using below mask
[1,0,0,0,0,0,0,0,0]
[1,1,0,0,0,0,0,0,0]
[1,1,1,0,0,0,0,0,0]
[1,1,1,1,0,0,0,0,0]
and then
[0,0,0,0,0,1,0,0,0]
[0,0,0,0,0,1,1,0,0]
[0,0,0,0,0,1,1,1,0]
[0,0,0,0,0,1,1,1,1]

but I was having trouble implementing this. Is this possible?

**Comment by erfanzar** [comment]
i guess i should implement the llama in jax just like how i did that in mojo and search for a way to access to the pointers of array and use jax.lax.fori_loop convert loop to xla for more speed up.

but yes it's possible to pad the array like that divide the array cols by 2 and copy half of that and concatenate that with zeros array
